thecodingmachine/gotenberg
# metrics :: add prometheus metrics

**Summary**



This PR implements the following **features**

* Prometheus Metrics
* Custom metrics on current number of chrome rendering 



Hi team, 

First, thanks for this amazing work, the project rocks!

The project can be more awesome, if we could implement more metrics to facilitate the monitoring and scaling of the service under load, in fact besides the CPU et RAM usage, the project doesn't provide enough metrics to ensure a more efficient scaling under a big load.

I've added these custom metrics to have a more reliable and convenient metric on which we can scale the Gotenberg service when is deployed on solutions like Kubernetes.

It's more convenient to scale service as Gotenberg based on his current state rather than the current resources usage because each render is different from another intern of resource usage and obviously time rendering, with this custom metric we can imagine implementing an HPA (HorizontalPodAutoscaler) which will scale up /down our service based on the current number of rendering and have a more efficient way to determine if our current stack can handle more requests or it has to be scaled up. 

Of course, my code is not perfect, and if you have any advice or want to discuss this usage, feel free to ping me!

If this PR and feature seem legit to you, I will continue and update the documentation adding a metrics section.



**Test plan (required)**

How to test?
* Run a K6 loadtest to convert an URL to PDF.
* Run this command to get the metrics:  `curl 127.0.0.1:3000/metrics`
* At this point you should see all the default metrics of Prometheus as well as our new custom one named `chrome_current_rendering`
 
**Checklist**

- [x] Have you followed the guidelines in our [CONTRIBUTING](CONTRIBUTING.md) guide?
- [x] Have you lint your code locally prior to submission (`make lint`)?
- [x] Have you successfully ran tests with your changes locally (`make tests`)?
- [ ] Have you updated the documentation (Markdown files under `build > docs > content` and then `make doc`)?
- [ ] I have squashed any insignificant commits
